src/v/cluster/persisted_stm.cc

```diff
@@ -217,13 +195,23 @@ ss::future<bool> persisted_stm::wait_no_throw(
 }
 
 ss::future<> persisted_stm::start() {
-    auto maybe_reader = co_await _snapshot_mgr.open_snapshot();
-    if (maybe_reader) {
-        storage::snapshot_reader& reader = *maybe_reader;
-        co_await hydrate_snapshot(reader);
+    std::optional<stm_snapshot> maybe_snapshot;
+
+    try {
+        maybe_snapshot = co_await load_snapshot();
+    } catch (...) {
+        vassert(
+          false,
+          "Can't load snapshot from '{}'. Got error: {}",
+          _snapshot_mgr.snapshot_path(),
+          std::current_exception());
+    }
```

Comment:
please write commit messages that reflect the changes being made. this commit does far more than a 1-1 translation to co-routines indicated by its commit message.

for example, before a best_effort recovery policy didn't hard-stop the system, but now it appears that that is the case. why? `::start` didn't seem to apply the snapshot (even via the renamed hydrate_snapshot) but now it does.

Reply:
Split the commits